### PR TITLE
Assessment App By Chong Yuan

### PR DESCRIPTION
**Pull request by BillyYccc**
Hi! I have finished implementing the basic features of the assessment app,so I'm creating a pull request to this repo.
I will map the commit to the corresponding feature, here is the features list:

- [x] 1. setup the database : https://github.com/tsegismont/vertx-gsoc-2018-qotd/commit/15cebf8ba53e3d9ba7067c4f7f49ca5609f1cdad
- [x] 2. setup the web server & basic retrieval : https://github.com/tsegismont/vertx-gsoc-2018-qotd/commit/c3f110f064c8637cd9a5b314aa6526956f75d8a7
- [x] 3. post a new quote : https://github.com/tsegismont/vertx-gsoc-2018-qotd/commit/7c0a849052094a68b97cdc6e822d4bfb8f28713b
- [x] 4. realtime quotes : https://github.com/tsegismont/vertx-gsoc-2018-qotd/commit/49692f6dd916ede5eab5e2fa125eaf4b4b47eed0

In addition, I create some commits for minor improvements, please let me know if it's necessary to squash. I also set up a Travis-CI, here is the [link](https://travis-ci.org/BillyYccc/vertx-gsoc-2018-qotd).

### CODE REVIEW

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 1]
```diff
@@ -2,39 +2,66 @@
...
+
+    // router.route().handler(BodyHandler.create());
+
+    // router.route().consumes("application/json").produces("application/json");
+
+    router.get("/quotes").handler(this::getAllQuotes);
+
+    // setup database and web server
+
+    prepareDatabaseFuture.compose(v -> {
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: :+1:

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 2]
```diff
@@ -3,50 +3,38 @@
...
     // TODO setup database and web server and eventually...
-    startFuture.complete();
-  }
-
-  private Future<Void> runScript(String script) {
-    Future<Void> future = Future.future();
-    jdbcClient.getConnection(getConn -> {
-      if (getConn.succeeded()) {
-        SQLConnection connection = getConn.result();
-        connection.execute("RUNSCRIPT FROM '" + script + "'", exec -> {
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java:38]: Why not keep the `RUNSCRIPT` command from H2? Works like a charm and no need to manually load the SQL

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 3]
```diff
@@ -2,39 +2,66 @@
...
   }
+
+  private void getAllQuotes(RoutingContext routingContext) {
+    databaseService.getAllQuotes(res -> {
+      if (res.succeeded()) {
+        routingContext.response().setStatusCode(200)
+                      .putHeader("Content-Type", "application/json; charset=utf-8")
+                      .end(res.result().toString());
+      } else {
+        routingContext.response().setStatusCode(404)
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Minor: if it fails the error code should be 500, not 404

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 4]
```diff
@@ -29,11 +31,12 @@ public void start(Future<Void> startFuture) throws Exception {
 
     Router router = Router.router(vertx);
 
-    // router.route().handler(BodyHandler.create());
+     router.route().handler(BodyHandler.create());
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Minor: you could restrict the body capture to `POST` requests

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 5]
```diff
@@ -64,4 +67,36 @@ private void getAllQuotes(RoutingContext routingContext) {
       }
     });
   }
+
+  private void postNewQuote(RoutingContext routingContext) {
+    JsonObject body = routingContext.getBodyAsJson();
+    String text = body.getString("text");
+    String author = body.getString("author");
+    if (text == null) {
+      routingContext.response().setStatusCode(404)
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Not the right status code

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java [hunk 6]
```diff
@@ -64,4 +67,36 @@ private void getAllQuotes(RoutingContext routingContext) {
...
+      .put("text", text)
+      .put("author", author);
+
+    databaseService.postNewQuote(newQuote, res -> {
+      if (res.succeeded()) {
+        routingContext.response().setStatusCode(200)
+          .putHeader("Content-Type", "application/json; charset=utf-8")
+          .end(newQuote.encodePrettily());
+      } else {
+        routingContext.response().setStatusCode(400)
```
Comment (tsegismont) [inline src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Not the right status code for a server error

File: src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java [hunk 7]
```diff
@@ -40,11 +46,101 @@ public void tearDown(TestContext testContext) {
...
+  @Test
+  public void testPostNewQuoteWithNoText(TestContext testContext) {
+    JsonObject newQuoteWithNoText = new JsonObject()
+      .put("author", "testAuthor3");
+
+    webClient.post("/quotes")
+      .putHeader("Content-Type", "application/json; charset=utf-8")
+      .as(BodyCodec.jsonObject())
+      .sendJsonObject(newQuoteWithNoText, testContext.asyncAssertSuccess(response -> {
+        testContext.assertEquals(404, response.statusCode(), response.bodyAsString());
```
Comment (tsegismont) [inline src/test/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticleTest.java]: Not the expected status code

**Comment by tsegismont** [comment]
@BillyYccc good. A few fixes are needed though

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 2 above)
Comment (BillyYccc) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java:38]: I did not know the RUNSCRIPT command before, done in https://github.com/tsegismont/vertx-gsoc-2018-qotd/pull/4/commits/db64ea673452da95ccb2cf6070a1a95a6424371d.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 3 above)
Comment (BillyYccc) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: See https://github.com/tsegismont/vertx-gsoc-2018-qotd/pull/4/commits/524ed3e43178a39c3bcc8e28ce7e714bba5bc4c4.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 4 above)
Comment (BillyYccc) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: It has been corrected in https://github.com/tsegismont/vertx-gsoc-2018-qotd/pull/4/commits/b52608764c1e761755d11a61264d24ef9911945b.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 5 above)
Comment (BillyYccc) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: It is stated to use 404 in the README, but I guess 400 is the expected one, WDYT?

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 6 above)
Comment (BillyYccc) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: Fixed in https://github.com/tsegismont/vertx-gsoc-2018-qotd/pull/4/commits/f3f7f82f89c26b8d9789314666166bc76a9aea0f.

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 5 above)
Comment (BillyYccc) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: I made a possible fix in https://github.com/tsegismont/vertx-gsoc-2018-qotd/pull/4/commits/a212f6f9a938ee9ee5650b77219a6051f54569b9.

**Comment by BillyYccc** [comment]
Thanks for your reviewing 👍 !
There is another issue I suddenly find in the project, all the quotes returned by the server to the client do not include the quote_id, maybe I should make another query and return the id as well?

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 2 above)
Comment (tsegismont) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java:38]: :+1:

**Comment by tsegismont** [comment]
>maybe I should make another query and return the id as well?

not that's fine, that's not important

File: src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java (hunk 5 above)
Comment (tsegismont) [reply src/main/java/io/vertx/gsoc2018/qotd/QuoteOfTheDayVerticle.java]: @BillyYccc nice catch. My bad. I will update the README

**Comment by BillyYccc** [comment]
Okay, is there anything more required to do?

**Comment by tsegismont** [comment]
@BillyYccc nope we're all good here

**Comment by BillyYccc** [comment]
Closing this since the assessment is done.

Congratulations and wish you success! @Sammers21

**Comment by Sammers21** [comment]
@BillyYccc, thank you!

**Comment by tsegismont** [comment]
Thank you @BillyYccc ! Your proposal was strong and choosing was hard. I hope you can propose again next year!

**Comment by BillyYccc** [comment]
Thanks for your kind words! @tsegismont

**Pull request closed without merging**